jax/experimental/export/_shape_poly.py

```diff
@@ -933,6 +985,119 @@ def cmp_sequence(s1, s2, elem_cmp) -> int:
   return 0
 
 
+class SymbolicScope:
+  """Indentifies a scope for symbolic expressions.
+
+  All symbolic expressions that interact (e.g., appear in the argument shapes
+  for one JAX function invocation, or are involved in arithmetic operations)
+  must be from the same scope and must share the same SymbolicScope object.
+
+  Holds the constraints on symbolic expressions.
+
+  See [the README](https://github.com/google/jax/blob/main/jax/experimental/jax2tf/README.md#user-specified-symbolic-constraints)
+  for more details.
+
+  Args:
+    constraints_str: A sequence of constraints on symbolic dimension expressions,
+      of the form `e1 >= e2` or `e1 <= e2`.
+  """
+  def __init__(self,
+               constraints_str: Sequence[str] = ()):
+    self._location_frame = source_info_util.user_frame(source_info_util.current())
+    self._explicit_constraints: list[tuple[_DimExpr, str]] = []
+
+    # Keep an efficient representation of
+    # the explicit constraints for use during reasoning.
+    self._monomial_bounds: dict[_DimMon, tuple[float, float]] = {}
+    self._expr_bounds: dict[_DimExpr, tuple[float, float]] = {}
+
+    constraints = self._parse_constraints(constraints_str)
+    for c, c_str in zip(constraints, constraints_str):
+      if (const := c.to_constant()) is not None:
+        if const < 0:
+          raise ValueError(f"Unsatisfiable explicit constraint: {c_str}")
+        continue
+      self._explicit_constraints.append((c, c_str))
+      self._process_constraint(c, c_str)
+
+  def __str__(self) -> str:
+    extras = []
+    if self._explicit_constraints:
+      extras.append(" with constraints:")
+      for _, c_str in self._explicit_constraints:
+        extras.append(f"  {c_str}")
+    loc = source_info_util._summarize_frame(self._location_frame) if self._location_frame else "unknown"
+    return (
+        f"{id(self)} created at {loc}" +
+        "\n".join(extras))
+  __repr__ = __str__
+
+  def _parse_constraints(self, constraints_str: Sequence[str]) -> Sequence[_DimExpr]:
+    # Parse some contraints into the current scope.
+    def parse_one(cs: str) -> _DimExpr:
+      if not isinstance(cs, str):
+        raise ValueError(
+            f"symbolic_scope must be invoked with a string: got {repr(cs)}")
+      eq_pos = cs.find("=")
+      if eq_pos <= 0 or cs[eq_pos - 1] not in [">", "<"]:
+        raise ValueError("Constraint parsing error: must contain one of '>=' or '<='")
+      e1_str = cs[:eq_pos - 1]
+      e1, = _Parser(e1_str, None, repr(e1_str), self).parse()
+      e2_str = cs[eq_pos + 1:]
+      e2, = _Parser(e2_str, None, repr(e2_str), self).parse()
+      diff = e1 - e2 if cs[eq_pos - 1] == ">" else e2 - e1
+      return _ensure_poly(diff, "symbolic_scope", self)
+
+    if isinstance(constraints_str, str):
+      raise ValueError(
```

Comment:
I want to catch a mistake that I think will be common, to pass a single constraint as a string, instead of a singleton tuple or list, for users that do not use pytype.